File: src/kand/analyzer.clj
(ns kand.analyzer
  (:refer-clojure :exclude [name])
  (:require [kand.type :refer :all]
            [clojure.string :refer [split]]
            [kand.parser :refer [parse]]
            [cats.core :as m]
            [cats.monad.either :refer :all])
  (:import [kand.type Application If Def Lambda Primitive Symbol Num True False Unit Quote Err Module Import Eval]))

(defmulti analyze type)

(defmulti execute (fn [v _ _] (type v)))

(defmethod execute Lambda [{:keys [params body]} args env]
  (let [bproc (analyze body)
        module (:name (:core/*module* env))
        [result _] (bproc (merge env (zipmap (map #(keyword module (:name %))
                                                  params)
                                             args)))]
    [result env]))

(defmethod execute Primitive [{:keys [f]} args env]
  [(apply f args) env])

(defmethod execute :default [v _ env]
  [(->Err (str v " can't execute")) env])

(defmethod analyze Application [{:keys [exps]}]
  (let [[operator & operands] exps
        fproc (analyze operator)
        aprocs (map analyze operands)]
    (fn analyze-application [env]
      (let [[app] (fproc env)
            args (map first (map (fn [aproc] (aproc env)) aprocs))
            errs (filter #(instance? Err %) args)]
        (if (empty? errs)
          (execute app args env)
          [errs env])))))

(defmethod analyze Symbol [sym]
  (let [[m n] (split (:name sym) #"/")
        module (when n m)
        sym-name (if n n m)]
    (fn analyze-symbol [env]
      (let [module (or module (:name (:core/*module* env)))]
        (if-let [value (get env (keyword module sym-name))]
          [value env]
          [(->Err (str "Can't find symbol " sym-name)) env])))))

(defmethod analyze Def [{:keys [name body]}]
  (let [bodyf (analyze body)]
    (fn analyze-def [env]
      (let [[result env] (bodyf env)
            module (:core/*module* env)]
        [(->Unit) (assoc env (keyword (:name module) (:name name)) result)]))))

(defmethod analyze If [{:keys [pred t f]}]
  (let [pproc (analyze pred)
        tproc (analyze t)
        fproc (analyze f)]
    (fn analyze-if [env]
      (let [[p] (pproc env)]
        (if (= (->True) p)
          (tproc env)
          (fproc env))))))

(defmethod analyze Quote [value]
  (fn analyze-quote [env]
    [(:val value) env]))

(defmethod analyze Module [sym]
  (fn analyze-module [env]
    [(->Unit) (assoc env :core/*module* (:name sym))]))

(defmethod analyze Import [module]
  (fn analyze-import [env]
    (let [code (slurp (str (-> module :module :name) ".knd"))]
      ((analyze (->Eval (->Str code))) env))))

(defmethod analyze Eval [code]
  (fn analyze-eval [env]
    (let [result (m/fmap #(reduce (fn [[_ env] exp]
                                    ((analyze exp) env))
                                  [nil env] %)
                         (parse (-> code :code :val)))]
      (if (right? result)
        [(first (:right result)) (second (:right result))]
        [(->Err (:left result)) env]))))

(defmethod analyze :default [exp]
  (fn analyze-default [env]
    [exp env]))

File: eta-kand/src/Kand/Analyzer.hs
module Kand.Analyzer ( Env(..)
                     , KandType(..)
                     , analyze ) where

import Kand.Parser
import Kand.Type
import Kand.Env

-- executeApplication :: Expr -> [Expr] -> Env -> (Expr, Env)

-- executeApplication (Lambda parameters body) args env =
--   bproc $ extendEnv parameters args env
--   where
--     bproc = analyze body

-- executeApplication (Primitive f) args env = (f args, env)

analyze :: Expr -> Env -> (Either String KandType, Env)

analyze (ExprList (ExprDef:(ExprSymbol name):expr:_)) = \env ->
  let (result, newEnv) = analyze expr env
  in
    case result of
      Left s -> (Left s, env)
      Right value -> (Right KandUnit, insertValue name value newEnv)
      
analyze (ExprSymbol name) = \env ->
  let value = lookupValue name env
  in case value of
    Just x -> (Right x, env)
    Nothing -> (Left ("Unable to resolve symbol: " ++ name ++ " in this context"), env)

analyze ExprTrue = \env -> (Right (KandBool True), env)

analyze ExprFalse = \env -> (Right (KandBool False), env)

analyze (ExprString s) = \env -> (Right (KandString s), env)

analyze (ExprNum n) = \env -> (Right (KandNumber n), env)

analyze expr = \env -> (Left ("Unable to resolve expression: " ++ show expr), env)

-- analyze (Application (operator:operands)) =
--   \env ->
--      let (app, _) = (fproc env)
--          args = map fst (map (\aproc -> aproc env) aprocs)
--      in executeApplication app args env
--   where
--     fproc = analyze operator
--     aprocs = map analyze operands

-- analyze (Sym name) = \env -> (lookupValue name env, env)

-- analyze (Def (Sym name) exp) = \env -> (Unit, insertValue name exp env)

-- analyze (If predicate consequent alternative) =
--   \env ->
--     let ((Boolean p), _) = (pproc env)
--     in if p then cproc env else aproc env
--   where
--     pproc = analyze predicate
--     cproc = analyze consequent
--     aproc = analyze alternative
  
-- analyze x = \env -> (x, env)
